File: internal/services/sqs/sigv4.go
package sqs

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strings"
)

const (
	sigV4Algorithm = "AWS4-HMAC-SHA256"
	sigV4Service   = "sqs"
)

type signatureError struct {
	code   string
	status int
}

func (e signatureError) Error() string {
	return e.code
}

func (s *Server) verifySignature(r *http.Request) error {
	if s.config.AuthMode == "" || strings.EqualFold(s.config.AuthMode, "relaxed") {
		return nil
	}
	auth := strings.TrimSpace(r.Header.Get("Authorization"))
	if auth == "" {
		return signatureError{code: "AccessDenied", status: http.StatusForbidden}
	}
	if !strings.HasPrefix(auth, sigV4Algorithm+" ") {
		return signatureError{code: "AuthorizationHeaderMalformed", status: http.StatusBadRequest}
	}
	values := parseAuthParams(strings.TrimPrefix(auth, sigV4Algorithm+" "))
	credential := values["Credential"]
	signedHeaders := values["SignedHeaders"]
	signature := values["Signature"]
	accessKey, dateStamp, region, service, ok := parseCredentialScope(credential)
	if !ok || signedHeaders == "" || signature == "" {
		return signatureError{code: "AuthorizationHeaderMalformed", status: http.StatusBadRequest}
	}
	if !s.validCredential(accessKey, region, service) {
		return signatureError{code: "InvalidAccessKeyId", status: http.StatusForbidden}
	}
	if r.Header.Get("x-amz-date") == "" {
		return signatureError{code: "AuthorizationHeaderMalformed", status: http.StatusBadRequest}
	}
	payloadHash := r.Header.Get("x-amz-content-sha256")
	if payloadHash == "" {
		payloadHash = "UNSIGNED-PAYLOAD"
	} else if err := verifyPayloadHash(r, payloadHash); err != nil {
		return err
	}
	expected := s.signatureForRequest(r, dateStamp, region, signedHeaders, payloadHash)
	if !hmac.Equal([]byte(signature), []byte(expected)) {
		return signatureError{code: "SignatureDoesNotMatch", status: http.StatusForbidden}
	}
	return nil
}

func signatureErrorDetails(err error) (string, int) {
	if sigErr, ok := err.(signatureError); ok {
		return sigErr.code, sigErr.status
	}
	return "AccessDenied", http.StatusForbidden
}

func (s *Server) validCredential(accessKey string, region string, service string) bool {
	return accessKey == defaultString(s.config.AccessKeyID, "dev") &&
		region == defaultString(s.config.Region, "us-east-1") &&
		service == sigV4Service
}

func verifyPayloadHash(r *http.Request, payloadHash string) error {
	if payloadHash == "UNSIGNED-PAYLOAD" {
		return nil
	}
	if strings.HasPrefix(payloadHash, "STREAMING-") {
		return signatureError{code: "NotImplemented", status: http.StatusNotImplemented}
	}
	body, err := io.ReadAll(r.Body)
	if err != nil {
		return signatureError{code: "InvalidRequest", status: http.StatusBadRequest}
	}
	_ = r.Body.Close()
	r.Body = io.NopCloser(bytes.NewReader(body))
	if got := sha256Hex(body); !hmac.Equal([]byte(strings.ToLower(payloadHash)), []byte(got)) {
		return signatureError{code: "SignatureDoesNotMatch", status: http.StatusForbidden}
	}
	return nil
}

func (s *Server) signatureForRequest(r *http.Request, dateStamp string, region string, signedHeaders string, payloadHash string) string {
	canonicalRequest := strings.Join([]string{
		r.Method,
		canonicalURI(r.URL.Path),
		canonicalQueryString(r.URL.Query()),
		canonicalHeaders(r, signedHeaders),
		strings.ToLower(signedHeaders),
		payloadHash,
	}, "\n")
	amzDate := r.Header.Get("x-amz-date")
	scope := strings.Join([]string{dateStamp, region, sigV4Service, "aws4_request"}, "/")
	stringToSign := strings.Join([]string{
		sigV4Algorithm,
		amzDate,
		scope,
		sha256Hex([]byte(canonicalRequest)),
	}, "\n")
	signature := hmacSHA256(deriveSigningKey(s.config.SecretAccessKey, dateStamp, region), stringToSign)
	return hex.EncodeToString(signature)
}

func parseCredentialScope(credential string) (accessKey string, dateStamp string, region string, service string, ok bool) {
	parts := strings.Split(credential, "/")
	if len(parts) != 5 || parts[4] != "aws4_request" {
		return "", "", "", "", false
	}
	return parts[0], parts[1], parts[2], parts[3], true
}

func parseAuthParams(value string) map[string]string {
	result := map[string]string{}
	for _, part := range strings.Split(value, ",") {
		key, val, ok := strings.Cut(strings.TrimSpace(part), "=")
		if ok {
			result[key] = val
		}
	}
	return result
}

func canonicalURI(path string) string {
	if path == "" {
		return "/"
	}
	return awsPercentEncode(path, "/~")
}

func canonicalQueryString(values url.Values) string {
	type pair struct {
		key   string
		value string
	}
	pairs := []pair{}
	for key, vals := range values {
		if len(vals) == 0 {
			pairs = append(pairs, pair{key: key})
			continue
		}
		for _, val := range vals {
			pairs = append(pairs, pair{key: key, value: val})
		}
	}
	sort.Slice(pairs, func(i, j int) bool {
		if pairs[i].key == pairs[j].key {
			return pairs[i].value < pairs[j].value
		}
		return pairs[i].key < pairs[j].key
	})
	encoded := make([]string, 0, len(pairs))
	for _, item := range pairs {
		encoded = append(encoded, awsPercentEncode(item.key, "~-_")+"="+awsPercentEncode(item.value, "~-_"))
	}
	return strings.Join(encoded, "&")
}

func canonicalHeaders(r *http.Request, signedHeaders string) string {
	var b strings.Builder
	for _, name := range strings.Split(strings.ToLower(signedHeaders), ";") {
		name = strings.TrimSpace(name)
		if name == "" {
			continue
		}
		value := r.Header.Get(name)
		if name == "host" {
			value = r.Host
		}
		b.WriteString(name)
		b.WriteByte(':')
		b.WriteString(normalizeHeaderValue(value))
		b.WriteByte('\n')
	}
	return b.String()
}

func normalizeHeaderValue(value string) string {
	return strings.Join(strings.Fields(value), " ")
}

func awsPercentEncode(value string, safe string) string {
	var b strings.Builder
	for i := 0; i < len(value); i++ {
		ch := value[i]
		if (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') || ch == '-' || ch == '_' || ch == '.' || ch == '~' || strings.ContainsRune(safe, rune(ch)) {
			b.WriteByte(ch)
			continue
		}
		b.WriteString("%")
		b.WriteString(strings.ToUpper(hex.EncodeToString([]byte{ch})))
	}
	return b.String()
}

func deriveSigningKey(secret string, dateStamp string, region string) []byte {
	secret = defaultString(secret, "dev")
	dateKey := hmacSHA256([]byte("AWS4"+secret), dateStamp)
	regionKey := hmacSHA256(dateKey, region)
	serviceKey := hmacSHA256(regionKey, sigV4Service)
	return hmacSHA256(serviceKey, "aws4_request")
}

func hmacSHA256(key []byte, value string) []byte {
	mac := hmac.New(sha256.New, key)
	_, _ = mac.Write([]byte(value))
	return mac.Sum(nil)
}

func sha256Hex(value []byte) string {
	sum := sha256.Sum256(value)
	return hex.EncodeToString(sum[:])
}

File: services/sqs/src/sigv4.rs
//! Mirrors `internal/services/sqs/sigv4.rs`.
//!
//! AWS SigV4 verification for the SQS service (`service = "sqs"`), with the same
//! error codes + HTTP statuses the legacy server returns. Relaxed mode (default)
//! accepts everything; strict mode performs a full canonical-request signature
//! comparison. The signer is exposed (`signature_for_request`) so tests can sign
//! a request and assert round-trip acceptance.

use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256};

type HmacSha256 = Hmac<Sha256>;

pub const SIGV4_ALGORITHM: &str = "AWS4-HMAC-SHA256";
pub const SIGV4_SERVICE: &str = "sqs";

/// A signature failure carrying the AWS error code + HTTP status from the legacy
/// server.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SignatureError {
    pub code: &'static str,
    pub status: u16,
}

fn err(code: &'static str, status: u16) -> SignatureError {
    SignatureError { code, status }
}

/// Minimal request view the verifier needs.
pub struct SignedRequest<'a> {
    pub method: &'a str,
    pub path: &'a str,
    /// Raw query string (`a=1&b=2`), as received.
    pub query: &'a str,
    pub host: &'a str,
    pub authorization: &'a str,
    pub amz_date: &'a str,
    pub content_sha256: &'a str,
    /// Resolves a header value by lowercase name.
    pub header: &'a dyn Fn(&str) -> Option<String>,
    pub body: &'a [u8],
}

pub struct Credentials<'a> {
    pub auth_mode: &'a str,
    pub access_key_id: &'a str,
    pub secret_access_key: &'a str,
    pub region: &'a str,
}

fn default_str<'a>(value: &'a str, fallback: &'a str) -> &'a str {
    if value.is_empty() {
        fallback
    } else {
        value
    }
}

/// Mirrors `Server.verifySignature`.
pub fn verify_signature(req: &SignedRequest, creds: &Credentials) -> Result<(), SignatureError> {
    if creds.auth_mode.is_empty() || creds.auth_mode.eq_ignore_ascii_case("relaxed") {
        return Ok(());
    }
    let auth = req.authorization.trim();
    if auth.is_empty() {
        return Err(err("AccessDenied", 403));
    }
    let prefix = format!("{SIGV4_ALGORITHM} ");
    if !auth.starts_with(&prefix) {
        return Err(err("AuthorizationHeaderMalformed", 400));
    }
    let values = parse_auth_params(&auth[prefix.len()..]);
    let credential = values.get("Credential").cloned().unwrap_or_default();
    let signed_headers = values.get("SignedHeaders").cloned().unwrap_or_default();
    let signature = values.get("Signature").cloned().unwrap_or_default();
    let (access_key, date_stamp, region, service) = match parse_credential_scope(&credential) {
        Some(v) => v,
        None => return Err(err("AuthorizationHeaderMalformed", 400)),
    };
    if signed_headers.is_empty() || signature.is_empty() {
        return Err(err("AuthorizationHeaderMalformed", 400));
    }
    if !valid_credential(&access_key, &region, &service, creds) {
        return Err(err("InvalidAccessKeyId", 403));
    }
    if req.amz_date.is_empty() {
        return Err(err("AuthorizationHeaderMalformed", 400));
    }
    let payload_hash = if req.content_sha256.is_empty() {
        "UNSIGNED-PAYLOAD".to_string()
    } else {
        verify_payload_hash(req, req.content_sha256)?;
        req.content_sha256.to_string()
    };
    let expected = signature_for_request(
        req,
        &date_stamp,
        &region,
        &signed_headers,
        &payload_hash,
        creds,
    );
    if !constant_time_eq(signature.as_bytes(), expected.as_bytes()) {
        return Err(err("SignatureDoesNotMatch", 403));
    }
    Ok(())
}

fn valid_credential(access_key: &str, region: &str, service: &str, creds: &Credentials) -> bool {
    access_key == default_str(creds.access_key_id, "dev")
        && region == default_str(creds.region, "us-east-1")
        && service == SIGV4_SERVICE
}

fn verify_payload_hash(req: &SignedRequest, payload_hash: &str) -> Result<(), SignatureError> {
    if payload_hash == "UNSIGNED-PAYLOAD" {
        return Ok(());
    }
    if payload_hash.starts_with("STREAMING-") {
        return Err(err("NotImplemented", 501));
    }
    let got = sha256_hex(req.body);
    if !constant_time_eq(payload_hash.to_ascii_lowercase().as_bytes(), got.as_bytes()) {
        return Err(err("SignatureDoesNotMatch", 403));
    }
    Ok(())
}

/// Computes the SigV4 hex signature for a request. Public so tests can sign.
pub fn signature_for_request(
    req: &SignedRequest,
    date_stamp: &str,
    region: &str,
    signed_headers: &str,
    payload_hash: &str,
    creds: &Credentials,
) -> String {
    let canonical_request = [
        req.method,
        &canonical_uri(req.path),
        &canonical_query_string(req.query),
        &canonical_headers(req, signed_headers),
        &signed_headers.to_ascii_lowercase(),
        payload_hash,
    ]
    .join("\n");
    let scope = [date_stamp, region, SIGV4_SERVICE, "aws4_request"].join("/");
    let string_to_sign = [
        SIGV4_ALGORITHM,
        req.amz_date,
        &scope,
        &sha256_hex(canonical_request.as_bytes()),
    ]
    .join("\n");
    let key = derive_signing_key(creds.secret_access_key, date_stamp, region);
    hex_lower(&hmac_sha256(&key, string_to_sign.as_bytes()))
}

fn parse_credential_scope(credential: &str) -> Option<(String, String, String, String)> {
    let parts: Vec<&str> = credential.split('/').collect();
    if parts.len() != 5 || parts[4] != "aws4_request" {
        return None;
    }
    Some((
        parts[0].to_string(),
        parts[1].to_string(),
        parts[2].to_string(),
        parts[3].to_string(),
    ))
}

fn parse_auth_params(value: &str) -> std::collections::HashMap<String, String> {
    let mut result = std::collections::HashMap::new();
    for part in value.split(',') {
        if let Some((k, v)) = part.trim().split_once('=') {
            result.insert(k.to_string(), v.to_string());
        }
    }
    result
}

fn canonical_uri(path: &str) -> String {
    if path.is_empty() {
        return "/".to_string();
    }
    aws_percent_encode(path, "/~")
}

/// Mirrors `canonicalQueryString`: decode each `k=v`, sort by (key, value),
/// re-encode. Operates on the raw query string.
pub fn canonical_query_string(query: &str) -> String {
    if query.is_empty() {
        return String::new();
    }
    let mut pairs: Vec<(String, String)> = Vec::new();
    for item in query.split('&') {
        match item.split_once('=') {
            Some((k, v)) => pairs.push((url_decode(k), url_decode(v))),
            None => pairs.push((url_decode(item), String::new())),
        }
    }
    pairs.sort_by(|a, b| a.0.cmp(&b.0).then(a.1.cmp(&b.1)));
    pairs
        .iter()
        .map(|(k, v)| {
            format!(
                "{}={}",
                aws_percent_encode(k, "~-_"),
                aws_percent_encode(v, "~-_")
            )
        })
        .collect::<Vec<_>>()
        .join("&")
}

fn canonical_headers(req: &SignedRequest, signed_headers: &str) -> String {
    let mut out = String::new();
    for name in signed_headers.to_ascii_lowercase().split(';') {
        let name = name.trim();
        if name.is_empty() {
            continue;
        }
        let value = if name == "host" {
            req.host.to_string()
        } else {
            (req.header)(name).unwrap_or_default()
        };
        out.push_str(name);
        out.push(':');
        out.push_str(&normalize_header_value(&value));
        out.push('\n');
    }
    out
}

/// Mirrors `normalizeHeaderValue`: collapse runs of whitespace to single spaces.
pub fn normalize_header_value(value: &str) -> String {
    value.split_whitespace().collect::<Vec<_>>().join(" ")
}

/// Mirrors `awsPercentEncode`.
pub fn aws_percent_encode(value: &str, safe: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for &c in value.as_bytes() {
        let ch = c as char;
        if ch.is_ascii_alphanumeric()
            || ch == '-'
            || ch == '_'
            || ch == '.'
            || ch == '~'
            || safe.contains(ch)
        {
            out.push(ch);
        } else {
            out.push('%');
            out.push_str(&format!("{:02X}", c));
        }
    }
    out
}

fn url_decode(s: &str) -> String {
    let bytes = s.as_bytes();
    let mut out = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        match bytes[i] {
            b'%' if i + 2 < bytes.len() => {
                if let Ok(b) = u8::from_str_radix(&s[i + 1..i + 3], 16) {
                    out.push(b);
                    i += 3;
                    continue;
                }
                out.push(bytes[i]);
                i += 1;
            }
            b'+' => {
                out.push(b' ');
                i += 1;
            }
            b => {
                out.push(b);
                i += 1;
            }
        }
    }
    String::from_utf8_lossy(&out).into_owned()
}

fn derive_signing_key(secret: &str, date_stamp: &str, region: &str) -> Vec<u8> {
    let secret = default_str(secret, "dev");
    let date_key = hmac_sha256(format!("AWS4{secret}").as_bytes(), date_stamp.as_bytes());
    let region_key = hmac_sha256(&date_key, region.as_bytes());
    let service_key = hmac_sha256(&region_key, SIGV4_SERVICE.as_bytes());
    hmac_sha256(&service_key, b"aws4_request")
}

fn hmac_sha256(key: &[u8], value: &[u8]) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC accepts any key length");
    mac.update(value);
    mac.finalize().into_bytes().to_vec()
}

/// Lowercase hex SHA-256. Public for the canonical-payload tests.
pub fn sha256_hex(value: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(value);
    hex_lower(&hasher.finalize())
}

fn hex_lower(bytes: &[u8]) -> String {
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        s.push_str(&format!("{:02x}", b));
    }
    s
}

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}
